web/src/stream_list.ts

```diff
@@ -268,88 +268,67 @@ export function build_stream_list(force_rerender: boolean): void {
         return;
     }
 
-    const $parent = $("#stream_filters");
-    const elems = [];
-
-    function add_sidebar_li(stream_id: number): void {
+    function add_sidebar_li(stream_id: number, $list: JQuery): void {
         const sidebar_row = stream_sidebar.get_row(stream_id);
         assert(sidebar_row !== undefined);
         sidebar_row.update_whether_active();
-        elems.push(sidebar_row.get_li());
+        const $li = sidebar_row.get_li();
+        $list.append($li);
     }
 
     clear_topics();
-    $parent.empty();
-
-    const any_pinned_streams =
-        stream_groups.pinned_streams.length > 0 || stream_groups.muted_pinned_streams.length > 0;
-    const any_normal_streams =
-        stream_groups.normal_streams.length > 0 || stream_groups.muted_active_streams.length > 0;
-    const any_dormant_streams = stream_groups.dormant_streams.length > 0;
-
-    const need_section_subheaders =
-        (any_pinned_streams ? 1 : 0) +
-            (any_normal_streams ? 1 : 0) +
-            (any_dormant_streams ? 1 : 0) >=
-        2;
-
-    if (any_pinned_streams && need_section_subheaders) {
-        elems.push(
-            $(
-                render_stream_subheader({
-                    subheader_name: $t({
-                        defaultMessage: "Pinned",
-                    }),
-                }),
-            ),
-        );
-    }
-
-    for (const stream_id of stream_groups.pinned_streams) {
-        add_sidebar_li(stream_id);
-    }
-
-    for (const stream_id of stream_groups.muted_pinned_streams) {
-        add_sidebar_li(stream_id);
-    }
-
-    if (any_normal_streams && need_section_subheaders) {
-        elems.push(
+    $("#stream_filters").empty();
+    for (const section of stream_groups.sections) {
+        $("#stream_filters").append(
             $(
-                render_stream_subheader({
-                    subheader_name: $t({
-                        defaultMessage: "Active",
-                    }),
+                render_stream_list_section_container({
+                    id: section.id,
+                    section_title: section.section_title,
                 }),
             ),
         );
-    }
+        const is_empty = section.streams.length === 0 && section.muted_streams.length === 0;
+        $(`#stream-list-${section.id}-container`).toggleClass("no-display", is_empty);
 
-    for (const stream_id of stream_groups.normal_streams) {
-        add_sidebar_li(stream_id);
+        for (const stream_id of section.streams) {
+            add_sidebar_li(stream_id, $(`#stream-list-${section.id}`));
+        }
+        for (const stream_id of section.muted_streams) {
+            add_sidebar_li(stream_id, $(`#stream-list-${section.id}`));
+        }
     }
+    unread_ui.update_unread_counts();
```

Comment:
This is a big hammer that can be quite expensive; do we really need to call it here? It recomputes and rerenders unread counts everywhere in the UI, which mean every time we build this component we'll trigger rebuilding a bunch of other components too.

I would have thought this component would already be taking care of presenting accurate initial unread counts when built; I at least don't see a similar call pre-patch.

Reply:
Okay I figured it out -- we do need to do it because when we rerender the sidebar, we need to decide which counts to show and which to hide while rendering, which to mark as a muted count, etc. Either we should store all that data (which would be more complex) or we can update the counts again. I'm adding it back.

Reply:
I think there's likely a much smaller thing you can call. The right sidebar does not need a rerender, for example. 

The computation is  `unread.get_counts`, if that's what you need. But we surely already call that for rendering the left sidebar. So I think the issue is you just don't have the data flow right to use the existing call.

Reply:
OK. We should file an issue for 12.0 to look at optimizing how much of the view-change time can be optimized; I feel like the sidebars are now like 40% of the time and probably should not be.